certbot-ci/certbot_integration_tests/certbot_tests/assertions.py

```diff
@@ -0,0 +1,81 @@
+"""This module contains advanced assertions for the certbot integration tests."""
+import os
+import grp
+
+
+# Defining __all__ allows to make a static import "from assertions import *" without importing
+# also the modules that are part of this module.
+__all__ = ['assert_hook_execution', 'assert_save_renew_hook', 'assert_certs_count_for_lineage',
+           'assert_equals_permissions', 'assert_equals_group_owner', 'assert_world_permissions']
+
+
+def assert_hook_execution(probe_path, probe_content):
+    """
+    Assert that a certbot hook has been executed
+    :param probe_path: path to the file that received the hook output
+    :param probe_content: content expected when the hook is executed
+    """
+    with open(probe_path, 'r') as file:
+        lines = file.readlines()
+
+    # Comparing pattern to each line avoids to match "pre" for a line with "pre-override"
+    assert [line for line in lines if line == '{0}{1}'.format(probe_content, os.linesep)]
```

Comment:
Here and for the other related observations about the other assertions. Pytest reporting on assert is far superior to the standard unittest module. If this kind of assertion fails, it will display a complete report explaining why it is failing.

Here for instance, it will display the content of each array, with an emphase on the row where there is a difference. 

All the assertions here are properly registered to take advantage of that behavior.